nexus/deployment/src/planner.rs

```diff
@@ -0,0 +1,227 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! High-level facilities for generating Blueprints
+//!
+//! See crate-level documentation for details.
+
+use crate::blueprint_builder::BlueprintBuilder;
+use crate::blueprint_builder::Error;
+use nexus_types::deployment::Blueprint;
+use nexus_types::deployment::Policy;
+use slog::{info, Logger};
+
+pub struct Planner<'a> {
+    log: Logger,
+    policy: &'a Policy,
+    blueprint: BlueprintBuilder<'a>,
+}
+
+impl<'a> Planner<'a> {
+    pub fn new_based_on(
+        log: Logger,
+        parent_blueprint: &'a Blueprint,
+        policy: &'a Policy,
+        creator: &str,
+    ) -> Planner<'a> {
+        let blueprint =
+            BlueprintBuilder::new_based_on(parent_blueprint, policy, creator);
+        Planner { log, policy, blueprint }
+    }
+
+    pub fn plan(mut self) -> Result<Blueprint, Error> {
+        self.do_plan()?;
+        Ok(self.blueprint.build())
+    }
+
+    fn do_plan(&mut self) -> Result<(), Error> {
+        // The only thing this planner currently knows how to do is add services
+        // to a sled that's missing them.  So let's see if we're in that case.
+
+        // Internal DNS is a prerequisite for bringing up all other zones.  At
+        // this point, we assume that internal DNS (as a service) is already
+        // functioning.  At some point, this function will have to grow the
+        // ability to determine whether more internal DNS zones need to be
+        // added and where they should go.  And the blueprint builder will need
+        // to grow the ability to provision one.
+
+        for (sled_id, sled_info) in &self.policy.sleds {
+            // Check for an NTP zone.  Every sled should have one.  If it's not
+            // there, all we can do is provision that one zone.  We have to wait
+            // for that to succeed and synchronize the clock before we can
+            // provision anything else.
+            if self.blueprint.sled_ensure_zone_internal_ntp(*sled_id)? {
```

Comment:
Is this method _slightly_ misnamed because it returns true for boundary NTP too? Maybe just `sled_ensure_zone_ntp` would be more accurate? I assume we will eventually need a check for boundary NTP explicitly (to add more if needed), which might end up making this more confusing.

Reply:
Good point.  Changed in c58c7edae to `sled_ensure_zone_ntp`.  (It's a little weird: it's ensuring there's an NTP zone, but always adding an internal one.  I did consider having the planner add boundary ones if needed, but that's more complicated because of #4791.  But I agree this name is less misleading than what was there.)